src/type_checker/infer/call.rs

```diff
@@ -77,78 +75,25 @@ fn property_call(
-
-                    let mut raises = HashSet::new();
-                    let mut arg_names = vec![];
-                    let mut env = env.clone();
-                    for arg in args {
-                        let (arg_ty, new_env) = infer(arg, &env, ctx)?;
-                        arg_names.push(TypeName::from(&arg_ty.expr_ty(&arg.pos)?));
-                        raises = raises.union(&arg_ty.raises).cloned().collect();
-                        env = new_env;
-                    }
-
-                    let function = instance.fun(&name, &arg_names, &property.pos)?;
-                    let function_self_mut = function.self_mutable == Some(false);
-                    if !function_self_mut && mutable {
-                        let msg = format!("Cannot call {}, which expects self to be mutable", name);
-                        return Err(vec![TypeErr::new(&property.pos, &msg)]);
-                    }
-
-                    let function_ty_name = &function.ty().ok_or_else(|| {
-                        vec![TypeErr::new(&property.pos, "Cannot get type of function")]
-                    })?;
-                    (
-                        ctx.lookup(&function_ty_name, &property.pos)?,
-                        function_self_mut,
-                        function.raises
-                    )
-                }
-                _ =>
-                    return Err(vec![TypeErr::new(
-                        &inner_instance.pos,
-                        "Expected function or field"
-                    )]),
-            };
-
-            let (infer_ty, env) =
-                property_call(mutable && property_mutable, &instance, property, ctx, &env)?;
-            Ok((infer_ty.union_raises(&raises), env))
+            let ((mutable, inner_ty), env) =
+                property_call(mutable, instance, inner_instance, ctx, env)?;
+            let inner_expr = inner_ty.expr_ty(&inner_instance.pos)?;
+            property_call(mutable, &inner_expr, property, ctx, &env)
         }
 
-        _ => final_property_call(mutable, instance, property, ctx, env)
-    }
-}
-
-fn final_property_call(
-    mutable: bool,
-    instance: &ExpressionType,
-    property: &AST,
-    ctx: &Context,
-    env: &Environment
-) -> InferResult {
-    match &property.node {
         Node::Id { lit } => {
             let field = instance.field(&lit, &property.pos)?;
+            if field.private && field.in_class != env.state.in_class {
+                return Err(vec![TypeErr::new(&property.pos, &format!("{} is private", lit))]);
+            }
+
             let msg = format!("Cannot get type of field {}", field);
             let field_ty_name = &field.ty.ok_or_else(|| vec![TypeErr::new(&property.pos, &msg)])?;
-            Ok((InferType::from(&ctx.lookup(&field_ty_name, &property.pos)?), env.clone()))
+            let field_ty = InferType::from(&ctx.lookup(&field_ty_name, &property.pos)?);
+            Ok(((field.mutable, field_ty), env.clone()))
```

Comment:
Not sure why this is not caught by the tests, but this should probably be a conjunction of the field being mutable and current mutable (`field.mutable && mutable`)

Reply:
Realised this was slightly out of scope, but since this is so closely related I decided to fix it here and add tests for it here.